ansible_kernel/kernel.py

```diff
@@ -765,44 +846,81 @@ def get_module_doc(self, module):
         return data
 
     def is_ansible_alive(self):
-        if self.ansible_process is None:
+        if self.runner_thread is None:
+            logger.info("NOT STARTED")
             return False
-        return self.ansible_process.poll() is None
+        if self.runner_thread.is_alive():
+            logger.info("YES")
+        else:
+            logger.info("NO")
+        return self.runner_thread.is_alive()
 
-    def do_shutdown(self, restart):
+    def cancel_callback(self):
+        logger.info('called')
+        return self.shutdown_requested
 
-        logger = logging.getLogger('ansible_kernel.kernel.do_shutdown')
+    def finished_callback(self, runner):
+        logger.info('called')
+        self.shutdown = True
+        if not self.shutdown_requested:
+            self.queue.put(StatusMessage(['PlaybookEnded', {}]))
 
-        if self.ansible_process is None:
-            logger.debug("No ansible process")
-            return
+    def do_shutdown(self, restart):
 
-        try:
-            current_process = psutil.Process(self.ansible_process.pid)
-            children = current_process.children(recursive=True)
-            for child in children:
-                print('Child pid is {}'.format(child.pid))
-        except psutil.NoSuchProcess:
-            pass
 
         if self.is_ansible_alive():
-            logger.info('killing ansible {0}'.format(self.ansible_process.pid))
-            try:
-                self.ansible_process.kill()
-            except psutil.NoSuchProcess:
-                logger.info('process already dead {0}'.format(self.ansible_process.pid))
-            for child in children:
-                logger.info('killing ansible sub {0}'.format(child.pid))
-                try:
-                    child.kill()
-                except psutil.NoSuchProcess:
-                    logger.info('process already dead {0}'.format(child.pid))
-
-        logger.info('stopping helper')
-        self.helper.stop()
-        self.helper = None
-        self.tasks_counter = 0
+            self.shutdown = False
+            self.shutdown_requested = True
+
+            while not self.shutdown:
+                if not self.is_ansible_alive():
+                    break
+                logger.info("waiting for shutdown")
+                time.sleep(1)
+            logger.info("shutdown complete")
+
+        self.shutdown_requested = False
+        self.runner_thread = None
+        self.runner = None
+        if self.helper is not None:
+            self.helper.stop()
```

Comment:
This looks okay for shutting the thread down... do you know if Runner is able to kill the things correctly or does it leave orphans?

Reply:
I found some orphans this morning.   I'll do some more testing to see if this problem still occurs.  Jupyter Notebook doesn't give the kernel much time to shutdown before it kills it.